controllers/apps/operations/rebuild_instance.go

```diff
@@ -197,23 +230,284 @@ func (r rebuildInstanceOpsHandler) ReconcileAction(reqCtx intctrlutil.RequestCtx
+	// rebuild instances in place.
+	var (
+		completedCount int
+		failedCount    int
+	)
+	for i, instance := range rebuildInstance.Instances {
+		progressDetail := r.getInstanceProgressDetail(*compStatus, instance.Name)
+		if isCompletedProgressStatus(progressDetail.Status) {
+			completedCount += 1
+			if progressDetail.Status == appsv1alpha1.FailedProgressStatus {
+				failedCount += 1
+			}
+			continue
+		}
+		// rebuild instance
+		completed, err := r.rebuildInstanceInPlace(reqCtx, cli, opsRes, &progressDetail, rebuildInstance, instance, i)
+		if intctrlutil.IsTargetError(err, intctrlutil.ErrorTypeFatal) {
+			// If a fatal error occurs, this instance rebuilds failed.
+			progressDetail.SetStatusAndMessage(appsv1alpha1.FailedProgressStatus, err.Error())
+			setComponentStatusProgressDetail(opsRes.Recorder, opsRes.OpsRequest, &compStatus.ProgressDetails, progressDetail)
+			continue
+		}
+		if err != nil {
+			return 0, 0, err
+		}
+		if completed {
+			// if the pod has been rebuilt, set progressDetail phase to Succeed.
+			progressDetail.SetStatusAndMessage(appsv1alpha1.SucceedProgressStatus,
+				fmt.Sprintf("Rebuild pod %s successfully", instance.Name))
+		}
+		setComponentStatusProgressDetail(opsRes.Recorder, opsRes.OpsRequest, &compStatus.ProgressDetails, progressDetail)
+	}
+	return completedCount, failedCount, nil
+}
+
 // rebuildInstance rebuilds the instance.
-func (r rebuildInstanceOpsHandler) rebuildInstance(reqCtx intctrlutil.RequestCtx,
+func (r rebuildInstanceOpsHandler) rebuildInstanceInPlace(reqCtx intctrlutil.RequestCtx,
 	cli client.Client,
 	opsRes *OpsResource,
-	envForRestore []corev1.EnvVar,
 	progressDetail *appsv1alpha1.ProgressStatusDetail,
+	rebuildFrom appsv1alpha1.RebuildInstance,
 	instance appsv1alpha1.Instance,
-	backupName string,
 	index int) (bool, error) {
-	insHelper, err := r.prepareInstanceHelper(reqCtx, cli, opsRes, envForRestore, instance, backupName, index)
+	inPlaceHelper, err := r.prepareInplaceRebuildHelper(reqCtx, cli, opsRes, rebuildFrom.RestoreEnv,
+		instance, rebuildFrom.BackupName, index)
 	if err != nil {
 		return false, err
 	}
-	if backupName == "" {
-		return r.rebuildInstanceWithNoBackup(reqCtx, cli, opsRes, insHelper, progressDetail)
+
+	if rebuildFrom.BackupName == "" {
+		return inPlaceHelper.rebuildInstanceWithNoBackup(reqCtx, cli, opsRes, progressDetail)
+	}
+	return inPlaceHelper.rebuildInstanceWithBackup(reqCtx, cli, opsRes, progressDetail)
+}
+
+func (r rebuildInstanceOpsHandler) rebuildInstancesWithHScaling(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	opsRes *OpsResource,
+	rebuildInstance appsv1alpha1.RebuildInstance,
+	compStatus *appsv1alpha1.OpsRequestComponentStatus) (int, int, error) {
+	var (
+		completedCount int
+		failedCount    int
+		err            error
+	)
+	if len(compStatus.ProgressDetails) == 0 {
+		// 1. scale out the required instances
+		r.scaleOutRequiredInstances(opsRes, rebuildInstance, compStatus)
```

Comment:
Seems it would be more productive if the three steps could be chained as a workflow.

Reply:
This is just a comment... no need to modify this pr